homeassistant/components/fritz/camera.py

```diff
@@ -0,0 +1,126 @@
+import logging
+
+import pyqrcode
+
+from homeassistant.components.camera import Camera
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_UNKNOWN
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import dt as dt_util, slugify
+
+from .common import FritzBoxBaseEntity, FritzBoxTools
+from .const import (
+    DEFAULT_GUEST_WIFI_ENCRYPTION,
+    DEFAULT_GUEST_WIFI_QR_REFRESH_SEC,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Add a entities from a config_entry."""
+    _LOGGER.debug("Setting up FRITZ!Box camera entities")
+    fritzbox_tools: FritzBoxTools = hass.data[DOMAIN][entry.entry_id]
+
+    entities = await hass.async_add_executor_job(
+        _setup_guest_wifi_qr, fritzbox_tools, entry.title
+    )
+
+    async_add_entities(entities, True)
+
+
+def _setup_guest_wifi_qr(
+    fritzbox_tools: FritzBoxTools, device_friendly_name: str
+) -> list[FritzGuestWifiQRCamera]:
+    """Set up guest wifi entity to display the qr code."""
+    _LOGGER.debug("Setting up qr code camera entity")
+    ssid = fritzbox_tools.fritz_guest_wifi.ssid
+    return [
+        FritzGuestWifiQRCamera(
+            fritzbox_tools,
+            device_friendly_name,
+            ssid,
+        )
+    ]
+
+
+class FritzGuestWifiQRCamera(FritzBoxBaseEntity, Camera):
+    """Implementation of the FritzBox guest wifi QR code camera entity."""
+
+    QR_CODE = "QR code"
+
+    def __init__(
+        self,
+        fritzbox_tools: FritzBoxTools,
+        device_friendly_name: str,
+        ssid: str,
+    ) -> None:
+        """Initialize the camera."""
+        Camera.__init__(self)
+
+        self._attr_name = f"{device_friendly_name} {ssid} {self.QR_CODE}"
+        self._attr_unique_id = (
+            f"{fritzbox_tools.unique_id}-{slugify(ssid)}-{slugify(self.QR_CODE)}"
+        )
+        self._attr_icon = "mdi:qrcode-scan"
+        super().__init__(fritzbox_tools, device_friendly_name)
+
+        self.content_type = "image/svg+xml"
+        self._ssid: str = ssid
+        self._last_qr: bytes | None = None
+        self._refresh_at: datetime | None = None
+
+    @property
+    def extra_state_attributes(self) -> dict[str, str]:
+        """Return the attributes."""
+        attrs: dict[str, str] = {}
+        attrs["ssid"] = self._ssid or STATE_UNKNOWN
+        attrs["refresh_after_seconds"] = str(DEFAULT_GUEST_WIFI_QR_REFRESH_SEC)
+        return attrs
+
+    def _needs_refresh(self) -> bool:
+        """Check if a refresh of the qr code is necessary."""
+        if not (self._refresh_at and self._last_qr):
+            return True
+
+        return dt_util.utcnow() > self._refresh_at
+
+    def _generate_guest_wifi_qr(self) -> bytes:
+        """Generate a QR code for the guest wifi."""
```

Comment:
We don't allow device and service interfacing integrations to calculate state. That's only allowed for template integrations and similar integrations that rely on other integrations for state.

Device and service interfacing integrations should just set the state that is exported by their interface api.

Reply:
@MartinHjelmare so does that mean the call to pyqrcode needs to be in the external library and passed into HA here, eg as an image? Or does the camera feed need to be generated there too?

I hope my suggestion has not made this more complicated. My intention was the opposite 😊

Reply:
It's ok to have the library return a camera image when requested. I don't understand completely what you mean with camera feed in this context.

Reply:
I think what @MartinHjelmare means is to move _generate_guest_wifi_qr() into a separate class which belongs to common.py.

Did I understand correctly?

Reply:
No, we don't allow the integration to calculate the camera image. The camera image should come from the library api. If this is something that you want to add, the library needs to be updated to provide this image.

Reply:
Ok, now I understand. But if the project (fritzconnection) doesn't accept such a change (e.g. because there is a new requirement added (pyqrcode)) we could not add such a feature to Home Assistant? That's right?

Could we go the "service road" (original PR) in such a case or isn't this allowed either?

Reply:
I don't think we should create the QR code image in the integration, either in a camera entity or service call. The idea is that integrations should express the device or service api features and not invent new features that aren't natively supported by the device or service api.

Reply:
Ok. I will close this PR until such a functionality is ready in the underlying library (which I will try to implement).

Reply:
FYI I started a discussion at the `fritzconnection` project (https://github.com/kbr/fritzconnection/discussions/133)

Reply:
@MartinHjelmare kbr , the maintainer of `fritzconnection`, gave me the answer, that he don't itend to add such a functionality into `fritzconnection` because of the extra requirement of the  QR code lib (https://github.com/kbr/fritzconnection/discussions/133).

This means `fritzconnection` nor Home Assistant would like to implement such a functionality.  

Is there any chance to add this to HA except for a custom component?

Reply:
I think there may not be an obvious clean way to do this.  Here are some ideas.

1. Feels like a long shot but maybe the [existing QR code HA integration](https://www.home-assistant.io/integrations/qrcode/) could be somehow enhanced to not only read QR codes from video feeds but also generate qr codes from sensors, and expose them as an image somehow.  But that is quite a stretch and might be best as a new integration.
2. Maybe a command line component to create the QR.
3. Maybe a separate [docker container to generate QR codes on demand](https://hub.docker.com/r/samwierema/go-qr-generator).  Then you could use a jinja template to put a QR code on the HA user interface based on arbitrary text.  There might be online services that will do this for free but I couldn't easily find any.
4. If you have a working cgi-bin locally, QR codes can be generated directly via a cgi script.  [Here's a demo script I just made](https://github.com/davet2001/miscellaneous/blob/master/qr.cgi) - feel free to use it.

Reply:
My idea behind doing this is to use the service to set a new guest wifi password (which I contributed here: https://github.com/home-assistant/core/pull/62892) and afterwards generate a new QR code. Every 24h! Due to this PR here it uses the underlying library to get the new password. 

With the suggested ideas there is no way to direct get the password. This needs extra configuration of the credentials elsewhere. If this is the way to go I would prefer writing a custom component.

But what me make thinking about this is the statement from kbr in the `fritzconnection` discussion as he says:

> But from my point of view, you have done everything right: combine several libraries to build something new. Working with data from devices should happen at application level. I'm not familiar with HA, but I can imagine use cases where data are collected from devices and presented in some special way, out of the scope of the devices providing the data. So I'm wondering that HA does not provide a way to do this.

And I agree. I even thought HA should combine things and give new opportunities and not be limited by an underlying package....jm2c...

Reply:
Perhaps a good first step would be to add something that just exposed the password as a string.  Then a string to QR code converter could achieve what you are looking for without having to add lots to either fritzconnection or HA.

Reply:
I don't know if it's a good practice to save credentials as a clear text string and make it accessible through HA frontend.

Reply:
If that is a concern, a QR code would open the same security hole.

This is probably best discussed in the forum.

Reply:
But it's a little barrier because it's encrypted in an image. You must scan the code to retrieve the password.